File: python/set3/challenge23.py
from challenge21 import MersenneTwister


def untempering(p):
    """
    see https://occasionallycogent.com/inverting_the_mersenne_temper/index.html

    >>> mt = MersenneTwister(0)
    >>> mt.tempering(42)
    168040107
    >>> untempering(168040107)
    42
    """
    e = p ^ (p >> 18)
    e ^= (e << 15) & 0xEFC6_0000
    e ^= (e << 7) & 0x0000_1680
    e ^= (e << 7) & 0x000C_4000
    e ^= (e << 7) & 0x0D20_0000
    e ^= (e << 7) & 0x9000_0000
    e ^= (e >> 11) & 0xFFC0_0000
    e ^= (e >> 11) & 0x003F_F800
    e ^= (e >> 11) & 0x0000_07FF
    return e


def challenge23():
    ...


if __name__ == '__main__':
    challenge23()

File: kotlin/src/main/kotlin/io/github/ocirne/cryptopals/set3/Challenge23.kt
package io.github.ocirne.cryptopals.set3

import io.github.ocirne.cryptopals.set3.Challenge21.MersenneTwister

class Challenge23 {

    class Attacker {

        /** see https://occasionallycogent.com/inverting_the_mersenne_temper/index.html */
        fun untempering(p: UInt): UInt {
            var e = p xor (p shr 18)
            e = e xor ((e shl 15) and 0xEFC6_0000u)
            e = e xor ((e shl 7) and 0x0000_1680u)
            e = e xor ((e shl 7) and 0x000C_4000u)
            e = e xor ((e shl 7) and 0x0D20_0000u)
            e = e xor ((e shl 7) and 0x9000_0000u)
            e = e xor ((e shr 11) and 0xFFC0_0000u)
            e = e xor ((e shr 11) and 0x003F_F800u)
            e = e xor ((e shr 11) and 0x0000_07FFu)
            return e
        }

        fun replicateStream(origMt: MersenneTwister): MersenneTwister {
            val copyMt = MersenneTwister(0u)
            origMt.take(625).forEachIndexed { i, p ->
                copyMt.y[i] = untempering(p)
            }
            return copyMt
        }
    }
}
